sdk/core/core-amqp/src/retry.ts

```diff
@@ -194,3 +231,97 @@ export async function retry<T>(config: RetryConfig<T>): Promise<T> {
     throw lastError;
   }
 }
+
+/**
+ *
+ * Utility to exponentially retry on the given operation for until the specified maximum
+ * delay between operations is reached.
+ *
+ * @param {RetryConfig<T>} config Parameters to configure retry operation.
+ *
+ * @return {Promise<T>} Promise<T>.
+ */
+export async function exponentialRetry<T>(config: RetryConfig<T>): Promise<T> {
+  validateRetryConfig(config);
+  if (config.maxExponentialRetryDelayInMs == undefined || config.maxExponentialRetryDelayInMs < 0) {
+    config.maxExponentialRetryDelayInMs = defaultMaxDelayForExponentialRetryInMs;
+  }
+  if (config.minExponentialRetryDelayInMs == undefined || config.minExponentialRetryDelayInMs < 0) {
+    config.minExponentialRetryDelayInMs = defaultMinDelayForExponentialRetryInMs;
+  }
+
+  let lastError: MessagingError | undefined;
+  let result: any;
+  let success = false;
+
+  // Based on specified max delay, the total number of attempts to be made can be determined as,
+  // minExponentialRetryDelayInMilliseconds + 2^(retryAttempts) should be less than or equal to maxExponentialRetryDelayInMilliseconds
+  const totalNumberOfAttempts = Math.floor(
+    Math.log2(config.maxExponentialRetryDelayInMs - config.minExponentialRetryDelayInMs)
+  );
+
+  for (let i = 1; i <= totalNumberOfAttempts; i++) {
+    log.retry("[%s] Attempt number: %d", config.connectionId, config.operationType, i);
+    try {
+      result = await config.operation();
+      success = true;
+      log.retry(
+        "[%s] Success for '%s', after attempt number: %d.",
+        config.connectionId,
+        config.operationType,
+        i
+      );
+      if (result && !isDelivery(result)) {
+        log.retry(
+          "[%s] Success result for '%s': %O",
+          config.connectionId,
+          config.operationType,
+          result
+        );
+      }
+      break;
+    } catch (err) {
+      if (!err.translated) {
+        err = translate(err);
+      }
+
+      if (!err.retryable && err.name === "ServiceCommunicationError" && config.connectionHost) {
+        const isConnected = await checkNetworkConnection(config.connectionHost);
+        if (!isConnected) {
+          err.name = "ConnectionLostError";
+          err.retryable = true;
+        }
+      }
+      lastError = err;
+      log.error(
+        "[%s] Error occured for '%s' in attempt number %d: %O",
+        config.connectionId,
+        config.operationType,
+        i,
+        err
+      );
+
+      const targetDelayInMs = Math.min(
+        config.minExponentialRetryDelayInMs + Math.pow(2, i),
+        config.maxExponentialRetryDelayInMs
+      );
```

Comment:
With default values for min and max delays here, this results in a progression of 3, 5, 9, 17, 33, 65, 129, 257, 513, 1025, 2049 etc.

Questions:
- Shouldn't we be starting from the min value instead of min value + 2 ?
- This method differs from both what is in core-http and in storage. Which of the two did we take as reference here?
     - In the [exponential retry policy in storage](https://github.com/Azure/azure-sdk-for-js/blob/%40azure/storage-blob_12.0.0-preview.1/sdk/storage/storage-blob/src/policies/RetryPolicy.ts#L312-L317), `Math.pow(2, i)` is multiplied by the delay provided by the user instead of adding to it.
     - In the [exponential retry policy in core-http](https://github.com/Azure/azure-sdk-for-js/blob/%40azure/core-http_1.0.0-preview.1/sdk/core/core-http/lib/policies/exponentialRetryPolicy.ts#L131-L137), some other convoluted way is used

Reply:
> * Shouldn't we be starting from the min value instead of min value + 2 ?

This mirrors what we have in `core-http` and would depend on the configured value - did we want to set the default min to 0 ?
> * This method differs from both what is in core-http and in storage. Which of the two did we take as reference here?

We had decided to use what we have in `core-http` as captured in https://github.com/Azure/azure-sdk-for-js/issues/3331#issuecomment-508244667 as well, do we have consensus to go with one in `storage` instead? @bterlson @daviwil

Reply:
Overall, if we were to investigate best approach to take forward and not mirror `core-http` as decided, then following findings may help -

Based on discussion from cloud computing communities online,
> Slowing clients down may help, and the classic way to slow clients down is capped exponential backoff. Capped exponential backoff means that clients multiply their backoff by a constant after each attempt, up to some maximum value. In our case, after each unsuccessful attempt, clients sleep for:
![exponential-backoff-and-jitter-blog-figure-3](https://user-images.githubusercontent.com/45977823/60742890-f5eafa80-9f23-11e9-86dc-01fda8aaaaf9.png)

This seems to mirror what we have in `storage`. We'll likely need to update `core-http` as well.
And also decide if we want to do away with concept of minimum delay and use a maximum cap value along with retry attempts instead.

This would address, simplify other comment too where instead of `delayInSeconds` and min/max expoenential retry specific config, we can have one `maxRetryDelay` and `maxRetries` that would apply to both types of retry.
Additionally the number 2 can be made configurable as well, calling it the `backoffFactor`. (This seems to be a thing as well if you search for exponential time delay calculator /"online-backoff-calculator" online)

Reply:
_From https://github.com/Azure/azure-sdk-for-js/issues/3331#issuecomment-508244667_

> When an operation fails, the delay is computed as (2^retry-attempts) milliseconds and is thus incremented with each failed attempt in order of (1 ms, 3ms, 7ms, so on..) This is same as what we have in core-http

This is not what [core-http](https://github.com/Azure/azure-sdk-for-js/blob/%40azure/core-http_1.0.0-preview.1/sdk/core/core-http/lib/policies/exponentialRetryPolicy.ts#L131-L137) is doing.

In core-http, (2^retry-attempts) is not used as the time itself. This value is used to multiply with the delay (after introducing a randomization factor) provided by the user. Please review https://github.com/Azure/azure-sdk-for-js/blob/%40azure/core-http_1.0.0-preview.1/sdk/core/core-http/lib/policies/exponentialRetryPolicy.ts#L131-L137 one more time.

Reply:
Ah, yes missed the `*` in line 135. 
This makes it overall similar to what we have in `storage` excluding the use of minimum retry delay and boundedRandDelta computation.

Reply:
@ramya-rao-a @bterlson 
I've updated the implementation.
Moreover, @sadasant and I just discussed offline about having one copy of retry policy implementation where possible.

Should we house the exponential delay computation in one place and export, use that in here?
Would be good if we can clarify on which computation to use, so we can consider just having it modularized and exported, say from `core-http`.

Reply:
Actually, it isn't making sense to take on core-http as dependency. But eventually we could have a separate module for retry I think.

Reply:
Update: For clarity now made sure it's mirroring `core-http` much closely - @ramya-rao-a. Please check updated PR

Reply:
Latest update looks much better and closer to what core-http is doing.

Sharing the retry code between `core-http` and `core-amqp` is not trivial at the moment. `core-http` operates on the model of a "pipeline". The pipeline consists of "policies". Each policy does what it needs to do and then passes the "request" to the next policy in the pipeline. 

`core-amqp` does not have this concept of "pipeline". There are talks about making it have a pipeline, but that is not going to happen any time soon.

So, I am all in favor of aligning the retry logic between core-http and core-amqp, but not sharing actual re-usable code yet.